feat(form-core): add array method field.filterValues and form.filterFieldValues

Pascalmh opened the pull request:
### Problem

When setting a field back to an empty array 

```ts
const form = new FormApi({
  defaultValues: {
    people: [
        { firstName: "Bob" },
        { firstName: "Marry" },
    ],
  },
});

form.setFieldValue('people', [
  { firstName: "Marry" }, // <-- "Bob" should be removed
]);
```

`items[0].firstName` would still exist - when there's a validation on the firstName (on the Field-level) `form.canSubmit` would be false.

### Solution

adding `field.filterValues` and `form.filterFieldValues` to TSF:

```ts
form.filterFieldValues(
  'people',
  ({firstName}) => firstName === "Marry"
)
```

Pascalmh committed:
feat(form-core): add array method `field.filterValues` and `form.filterFieldValues`

Co-authored-by: LeCarbonator <18158911+LeCarbonator@users.noreply.github.com>

LeCarbonator commented:
I think if/when this gets merged, it would be great to rework the Arrays section of the docs to write about swap, insert, push, filter and clear. Thoughts? Should it be somewhere else?

LeCarbonator committed:
refactor(form-core): separate metaHelper function into multiple

The previous implementation of metaHelper could
lead to passing wrong index combinations with modes.
Splitting them into separate functions allows for explicit
requirements depending on mode.
The previous implementation is preserved for compatability, but
is marked as deprecated.

Pascalmh committed:
Merge remote-tracking branch 'upstream/main' into array-filterValues

LeCarbonator commented:
Oof, it did not like that refactor of metaHelper. I'll get to it later today.

LeCarbonator committed:
Merge branch 'main' of github.com:TanStack/form into array-filterValues

LeCarbonator committed:
chore: make filterFieldValues use DeepKeysOfType type
